src/auth/auth.service.ts

```diff
@@ -19,24 +22,94 @@ export class AuthService {
   }
 
   async authenticateGoogle(token: string) {
-    const tokenInfo = await this.oauthClient.getTokenInfo(token);
+    const { email } = await this.oauthClient.getTokenInfo(token);
+    const user = await this.userService.getUser(email);
+
+    if (user) {
+      return this.signIn(user);
+    }
+
+    return this.signUp(email);
+  }
+
+  async signUp(email: string) {
+    const newUser = await this.userService.create(email);
+    const tokens = await this.getTokens(newUser.id, newUser.email);
+
+    await this.userService.updateRefreshToken(
+      newUser.email,
+      await argon2.hash(tokens.refreshToken),
+    );
+
+    return tokens;
+  }
 
-    return tokenInfo;
+  async signIn(user: UserEntity) {
+    const tokens = await this.getTokens(user.id, user.email);
 
-    // const email = tokenInfo.email;
+    await this.userService.updateRefreshToken(
+      user.email,
+      await argon2.hash(tokens.refreshToken),
+    );
 
-    // try {
-    //   const user = await this.usersService.getByEmail(email);
+    return tokens;
+  }
+
+  async logout(email: string) {
+    return this.userService.updateRefreshToken(email, null);
+  }
+
+  async refreshTokens(email: string, refreshToken: string) {
+    const user = await this.userService.getUser(email);
 
-    //   return this.handleRegisteredUser(user);
-    // } catch (error) {
-    //   if (error.status !== 404) {
-    //     throw new error();
-    //   }
+    if (!user || !user.refreshToken) {
+      throw new ForbiddenException('Access Denied');
```

Comment:
Maybe extract into an AccessDeniedException?

Reply:
https://docs.nestjs.com/exception-filters#custom-exceptions

Reply:
Forbidden = access denied?

Reply:
> Forbidden = access denied?

Yep, but if u will make custom exception - it will be easier to reuse. Maybe it not so meaningul. I`ll just leave it on your own decision